merge enh/fitsstorage - Update DRAGONS to use Fits Storage 3.4.x for local calibration manager

phirstgemini opened the pull request:
* Updates localmanager.py to use new `FitsStorage` package and deprecates `GeminiObsDB` and `GeminiCalMgr`
* This changes the `SQLAlchemy` dependency from <2.0 to >=2.0.0 and allows use with the latest stable release.

Python environment changes needed to work with this are:
* Remove requirement to hold `SQLAlchemy` at <2.0, and update to latest stable `SQLAlchemy`
* Remove `GeminiObsDB` and `GeminiCalMgr` (uninstall not strictly required, they won't be used any more. Suggest uninstalling them as they require `SQLAlchemy<2.0` so you may end up in dependency jail if you don't)
* `pip install git+https://github.com/GeminiDRSoftware/FitsStorage.git@v3.4.0b1` - this is the 3.4.0 Beta 1 release. Obviously, a final release on the 3.4.x line will be coordinated prior to a DRAGONS release that uses this.

Note that #PR-475 is somewhat of a prerequisite for this - without PR-475 log messages from FitsStorage to the gempy logger will in many cases not come through correctly.

phirstgemini commented:
Not sure why it's complaining about CI. It shows green on jenkins.

KathleenLabrie commented:
Thanks for the installation instructions.  I'll need that to run the tutorial with the new stuff.  

Do the tests include things like `caldb config`, `caldb list`, `caldb add`, `caldb remove`.  (either command line or API)?

KathleenLabrie commented:
It just occurred to me that I believe most of the Jenkins DRAGONS tests do *not* use calibration association but explicitly state the calibration to use.

KathleenLabrie commented:
The first time I ran `caldb init -w` it took a very long time to run, like half a minute.  But it might have been some compiling of a newly installed software.   It seems more responsive now when I run it again.

When I first ran `caldb init -w`, I notice a very quick pop up window on my screen that disappeared right away.  No idea why there would be pop up window.

Otherwise, I ran the NIRI imaging tutorial (cmd line) and the behavior was normal.

No objections to merging.

teald commented on recipe_system/scripts/caldb.py:
```diff
@@ -137,18 +137,23 @@ def _action_remove(self, args):
+                       "existing database.")
+                print(msg)
             self._log("Initializing {}...".format(self.db.dbfile))
             self.db.init(wipe=args.wipe)
         except LocalManagerError as e:
             if e.error_type == ERROR_CANT_WIPE:
-                self.usage(message=msg)
+                print(f"Error wiping existing database file "
+                          f"{self.db.dbfile}")
             elif e.error_type == ERROR_CANT_CREATE:
-                log(e.message, sys.stderr, bold=True)
+                print(f"Error creating database file {self.db.dbfile}")
```
Should these print statements be log messages? I'm unsure what's used in this part of the code, if this is consistent with other parts of the cal service/recipe system then it's all good.

phirstgemini replied:
Honestly, it wasn't clear to me why the previous author used the combination of prints and logs. It seems they have it set up where things that the user needs to see interactively are prints, and the logs go to a file but not the console. I'm not sure why that is, but rather than changing it for the whole script, I just followed their usage...

teald commented on recipe_system/cal_service/localmanager.py:
```diff
@@ -164,25 +154,40 @@ def init_database(self, wipe=True):
+        if os.path.exists(db_path) and wipe:
+            try:
+                os.remove(db_path)
+            except Exception:
+                raise LocalManagerError(ERROR_CANT_WIPE,
+                                        "Failed to remove {db_path}")
```
```suggestion
            except Exception as err:
                raise LocalManagerError(ERROR_CANT_WIPE,
                                        "Failed to remove {db_path}") from err
```

phirstgemini replied:
This is in commit 564cdc654

phirstgemini commented:
> Thanks for the installation instructions. I'll need that to run the tutorial with the new stuff.
> 
> Do the tests include things like `caldb config`, `caldb list`, `caldb add`, `caldb remove`. (either command line or API)?

I didn't mess with those tests. There are tests that call `caldb.add_cal()`, `cal_db.remove_cal()` and `caldb.get_calibrations()` and they do pass with the new version. So my assumption was the tests passed before and they pass now and that indicates that I didn't break it. :-)

I did manually test the caldb script (and adjusted it for some differences between the old and new localmanager versions). I don't know how to / if it's possible or sensible to test the actual script from the "tests" (as opposed to the underlying calls that do the actual functionality, which are those mentioned above). 

One of the ghost tests did fail because it was getting the wrong calibration, which is a bug in this work that I fixed, so that reassures me that at least some of the tests do use the actual calibration association functionality.

phirstgemini commented on geminidr/ghost/calibration_ghost.py:
Yes, that's intentional. It actually should have been removed from the code base some time ago, I was surprised it was there and checked with Chris on slack:

PH> Am I correct in thinking that geminidr/ghost/calibration_ghost.py is not supposed to be being used by anything now?
CS> Yes you are. That file should have been removed from the codebase, my bad.

phirstgemini commented on recipe_system/cal_service/localmanager.py:
```diff
@@ -164,25 +154,40 @@ def init_database(self, wipe=True):
             self.session.commit()
         except OperationalError:
-            message = f"There was an error when trying to create the database {fsc.db_path}. "
-            message += "Please, check your path and permissions."
+            message = (f"There was an error when trying to create the database "
+                       f"{db_path}. Please, check your path and permissions.")
             raise LocalManagerError(ERROR_CANT_CREATE, message)
```
Thanks.. I wasn't really aware of that raise ... from ... construct, and apparently neither was the previous author. :-)
I'll change it as suggested

phirstgemini commented on recipe_system/cal_service/localmanager.py:
```diff
@@ -345,16 +334,14 @@ def calibration_search(self, rq, howmany=1, fullResult=False):
-            for res in query.order_by(File.name):
-                yield FileData(res[0], res[1])
+            headers = list_headers(selection, orderby, self.session)
         except OperationalError:
             message = "There was an error when trying to read from the database."
             raise LocalManagerError(ERROR_CANT_READ, message)
```
got it, thanks.

teald reviewed (approved):
lgtm!

KathleenLabrie reviewed (approved):
Tested with NIRI tutorial, using command line.  Everything worked.
